addons/mrp/report/mrp_report_bom_structure.py

```diff
@@ -218,7 +218,15 @@ def _get_bom_data(self, bom, warehouse, product=False, line_qty=False, bom_line=
                                                parent_product=product, index=new_index, product_info=product_info, ignore_stock=ignore_stock)
             else:
                 component = self._get_component_data(bom, product, warehouse, line, line_quantity, level + 1, new_index, product_info, ignore_stock)
-            components.append(component)
+
+            to_add = True
+            for component_bom in components:
+                if component['product_id'] == component_bom['product_id']:
+                    to_add = False
+                    self._merge_components(component_bom, component)
+
+            if to_add:
+                components.append(component)
```

Comment:
You could use a for else with a break in order to avoid browsing the list to the end.
```python
            for component_bom in components:
                if component['product_id'] == component_bom['product_id']:
                    self._merge_components(component_bom, component)
                    break
            else:
                components.append(component)